airbyte-workers/src/main/java/io/airbyte/workers/protocols/airbyte/StateDeltaTracker.java

```diff
@@ -0,0 +1,212 @@
+ */
+
+package io.airbyte.workers.protocols.airbyte;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * This class tracks "deltas" between states in compact {@code byte[]}s with the following schema:
+ *
+ * <pre>
+ *  [(state hash),(stream index),(record count)...] with the last two elements repeating per stream in the delta.
+ * </pre>
+ *
+ * This class also maintains a {@code Set} of {@code committedStateHashes} so that it can accumulate
+ * both committed and total record counts per stream.
+ * <p>
+ * The StateDeltaTracker is initialized with a memory limit. If this memory limit is exceeded, new
+ * states deltas will not be added and per-stream record counts will not be able to be computed.
+ * This is to prevent OutOfMemoryErrors from crashing the sync.
+ */
+@Slf4j
+public class StateDeltaTracker {
+
+  private static final int STATE_HASH_BYTES = Integer.BYTES;
+  private static final int STREAM_INDEX_BYTES = Short.BYTES;
+  private static final int RECORD_COUNT_BYTES = Long.BYTES;
+  private static final int BYTES_PER_STREAM = STREAM_INDEX_BYTES + RECORD_COUNT_BYTES;
+
+  private final ByteBuffer stateHashByteBuffer;
+  private final ByteBuffer streamIndexByteBuffer;
+  private final ByteBuffer recordCountByteBuffer;
+  private final Set<Integer> committedStateHashes;
+  private final HashMap<Short, Long> committedRecordsByStream;
+
+  /**
+   * Every time a state is added, a new byte[] containing the state hash and per-stream delta will be
+   * added to this list. Every time a state is committed, state deltas up to the committed state are
+   * removed from the head of the list and aggregated into the committed count map. The source thread
+   * adds while the destination thread removes, so those respective methods are synchronized to
+   * prevent thread-safety issues.
+   */
+  @VisibleForTesting
+  protected final List<byte[]> stateDeltas;
+
+  @VisibleForTesting
+  protected long remainingCapacity;
+  @VisibleForTesting
+  protected boolean capacityExceeded;
+
+  public StateDeltaTracker(final long memoryLimitBytes) {
+    this.stateHashByteBuffer = ByteBuffer.allocate(STATE_HASH_BYTES);
+    this.streamIndexByteBuffer = ByteBuffer.allocate(STREAM_INDEX_BYTES);
+    this.recordCountByteBuffer = ByteBuffer.allocate(RECORD_COUNT_BYTES);
+    this.committedStateHashes = new HashSet<>();
+    this.committedRecordsByStream = new HashMap<>();
+    this.stateDeltas = new ArrayList<>();
+    this.remainingCapacity = memoryLimitBytes;
+    this.capacityExceeded = false;
+  }
+
+  /**
+   * Converts the given state hash and per-stream record count map into a {@code byte[]} and stores
+   * it.
+   *
+   * This method is synchronized to provide thread safety between the source thread calling addState
```

Comment:
i don't think this synchronization gives you exactly the guarantees that you want. The only guarantee it gives is that `addState` won't be called at the same time by multiple callers. The guarantee that I think you would need for how this is structured now is that neither `addState` nor  `commitState` get called at the same time. 

I think you you can solve for this in one of two ways.
1. you can use the synchronized keyword slightly differently as described in this [section of this article](https://www.baeldung.com/java-synchronized#3-synchronized-blocks-within-methods). basically you're saying for any block that is synchronized on this object, only one code block can be executing at a time.
2. you can make sure every underlying instance variable that these methods are touching is a threadsafe data structure (e.g. `AtomicInteger`, `ConcurrentHashMap.newKeySet()`, etc). 

Happy to discuss it more! Concurrency is tricky :D

Reply:
Ended up going with option 1 because it seems to map better into my mental model for thinking about how these threads contend. 

I coded up a sandbox example with two threads simultaneously trying to add/remove from a list and verified that synchronizing the methods can lead to issues, while `synchronize (this)` blocks inside the methods gives the intended behavior, so hopefully my updates to the StateDeltaTracker look sound! Let me know how it looks to you